x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/AbstractPhysicalOperationProviders.java

```diff
@@ -274,16 +283,28 @@ private void aggregatesToFactory(
                     }
                     List<Integer> inputChannels = sourceAttr.stream().map(attr -> layout.get(attr.id()).channel()).toList();
                     assert inputChannels.stream().allMatch(i -> i >= 0) : inputChannels;
-                    if (aggregateFunction instanceof ToAggregator agg) {
-                        consumer.accept(new AggFunctionSupplierContext(agg.supplier(inputChannels), mode));
-                    } else {
-                        throw new EsqlIllegalArgumentException("aggregate functions must extend ToAggregator");
+
+                    AggregatorFunctionSupplier aggSupplier = null;
+                    aggSupplier = supplier(aggregateFunction, inputChannels);
+
+                    // apply the filter only in the initial phase - as the rest of the data is already filtered
+                    if (aggregateFunction.hasFilter() && mode.isInputPartial() == false) {
```

Comment:
This would make sense only if the filter is common across all aggregations so it can filter the data coming in across all aggs - see https://github.com/elastic/elasticsearch/issues/114352